### ISSUE TITLE
discussion: create System Addon section

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [ghacksuserjs/ghacks-user.js#171]
earthlng said
> btw how do you feel about adding a new section for all the system-addons prefs? There are likely more of them coming in the future. I recently updated an FF55nightly to 56 and it had ~~like 10 or 12~~ 14 system addons. Instead of adding some of the prefs to the Personal and others to the Quiet Fox section, I think it would be nice to have them all grouped together.

Thorin said
> /me nods

earthlng said
> adding experiments in there too is probably a good idea.
> In Nightly56 the following addons are included (there may be more that get installed later, idk!)
> [see checklist below]
> 
> from those I think e10s, aushelper and webcompat are probably of no concern.
> automigrate-bookmarks doesn't seem to be a system-addon (?)
> IDK about 0300. f.e screenshots is not about "quiet fox" at all, it only makes connections if the user initiates it. Maybe 1300? Or more towards the end where Personal stuff is, 2900 for example?
> 
> edit: 0300 is also pretty long already, better to use something else IMO

---

* [x] clicktoplay-rollout
<details><summary>clicktoplay-rollout discussion (click me)</summary><p>

@Thorin-Oakenpants wrote

> see [1365714 Release Flash plugin Click-to-Play (aka "Ask to Activate")](https://bugzilla.mozilla.org/show_bug.cgi?id=1365714) resolved fixed FF55 .. ~~prefs~~ string to look for contains `ctprollout`? There are none bar the addon itself. I think this is self contained - there are no prefs.
> Since we don't encourage Flash, and it would never deploy until Flash was present, we can ignore this.

@earthlng wrote

> > I think this is self contained - there are no prefs.
> 
> from looking at the [source](https://dxr.mozilla.org/mozilla-central/source/browser/extensions/clicktoplay-rollout) the addon consists of a single js file and it deals with `plugin.state.flash`. Default value for that, at least in ESR52, is `2` (=enabled), while the addon will set it to `1` ([=click-to-play](https://dxr.mozilla.org/mozilla-central/source/dom/plugins/base/nsIPluginTag.idl#16))
> I agree we can ignore this.

@earthlng wrote

> https://developer.mozilla.org/en-US/Firefox/Releases/55#Plugins
> > For Firefox 55 **release version**, the plan is to activate this for 5% of users 2 weeks after release, 25% of users 4 weeks after release, and **100% of users 6 weeks after release**
> 
> FF56 will probably have `plugin.state.flash` defaulting to `1` then and the addon will get trashed

</p></details>

* [x] presentation - `nothing to do`
* [x] aushelper - `nothing to do`
* [x] e10srollout - `nothing to do`
* [x] webcompat - `nothing to do`
* [x] webcompat-reporter - `0518` - https://github.com/ghacksuserjs/ghacks-user.js/commit/049841acb13f88dcc06fd5eef3ead4019000a346
* [x] activity-stream - `0514` - https://github.com/ghacksuserjs/ghacks-user.js/commit/71c81992f5607609746015e9f48661fed4cc45c7 (originally `0513`)
* [x] pocket - `0510` - https://github.com/ghacksuserjs/ghacks-user.js/commit/71c81992f5607609746015e9f48661fed4cc45c7
* [x] flyweb - `0511` - https://github.com/ghacksuserjs/ghacks-user.js/commit/71c81992f5607609746015e9f48661fed4cc45c7
* [x] formautofill - `0517` - https://github.com/ghacksuserjs/ghacks-user.js/commit/808196718736b84d9ec9938212c47fc23fa8fd87
* [x] onboarding - `0515` - https://github.com/ghacksuserjs/ghacks-user.js/commit/4e36051192b2ee20c0fc5a26473e8fc6f6f7b802
* [x] screenshots - `0516` - https://github.com/ghacksuserjs/ghacks-user.js/commit/16499e961100e750751fa9187e2b72174f736063
* [x] shield-recipe-client - `0512` - https://github.com/ghacksuserjs/ghacks-user.js/commit/bcf38ca80b0728d919ee87aecbe41b30d9bb8aef
* [x] follow-on-search - `0513` - https://github.com/ghacksuserjs/ghacks-user.js/commit/63ed779776d15440df93417c0204993a984bf591

...

**Comment by Thorin-Oakenpants** [comment]
can we self clean up our crap in diffs

I looked at 0300s and it's cramped. Hence I suggested 0399. I also considered your suggestion that it go towards the end, probably between the last section (ie the last section that deals with privacy/FPing/security etc), ie. 2800 SHUTDOWN and the personal section.

I also considered that we should do some section renumbering to leave gaps. eg shove personal to 8000, shove shutdown to 7000 ... leave some more room for future (eg 2600 is cramped, and we have had to resort to 2697, 2698, 2699)

Other points. auto-migrate I just added to the mix for consideration. I honestly believe we will never include it. I know screenshots is not about "quiet fox" and I didn't say that the proposed number 0399 would be called that, it is to be called SYSTEM ADDONS, so it fits, but it doesn't always fit the purpose of security/FPing/privacy etc. Hence I am glad you suggested elsewhere towards the end.

What to put in that section? That list includes autoformfill FFS. I do not like this fragmentation - its adding complexity :-(

Anyway
- Decision 1: where to put this, do we take this opportunity to space out some section numbers
- Decision 2: what to include

**Comment by Thorin-Oakenpants** [comment]
current sections

```
/*** 0100: STARTUP ***/
/*** 0200: GEOLOCATION ***/
/*** 0300: QUIET FOX
/*** 0400: BLOCKLISTS / SAFE BROWSING / TRACKING PROTECTION
/*** 0600: BLOCK IMPLICIT OUTBOUND [not explicitly asked for - eg clicked on] ***/
/*** 0800: LOCATION BAR / SEARCH BAR / SUGGESTIONS / HISTORY / FORMS [SETUP]
/*** 0900: PASSWORDS ***/
/*** 1000: CACHE [SETUP] ***/
/*** 1100: MULTI-PROCESS (e10s)
/*** 1200: HTTPS ( SSL/TLS / OCSP / CERTS / HSTS / HPKP / CIPHERS )
/*** 1400: FONTS ***/
/*** 1600: HEADERS / REFERERS [SETUP]
/*** 1700: CONTAINERS [SETUP]
/*** 1800: PLUGINS ***/
/*** 2000: MEDIA / CAMERA / MIC ***/
/*** 2200: UI MEDDLING
/*** 2300: WEB WORKERS [SETUP]
/*** 2400: DOM & JAVASCRIPT ***/
/*** 2500: HARDWARE FINGERPRINTING ***/
/*** 2600: MISC - LEAKS / FINGERPRINTING / PRIVACY / SECURITY ***/
/*** 2697: USER AGENT (UA) SPOOFING
/*** 2698: FIRST PARTY ISOLATION (FPI)
/*** 2699: privacy.resistFingerprinting
/*** 2700: COOKIES & DOM STORAGE ***/
/*** 2800: SHUTDOWN [SETUP]
/*** 3000: PERSONAL SETTINGS [SETUP]
/*** 9999: DEPRECATED / REMOVED / LEGACY / RENAMED
```

**Comment by 2glops** [comment]
where to put SYSTEM ADDONS : 
IMHO, before or after "PLUGINS", as they both extend the functions of the browser (even if plugins are dying)

**Comment by Thorin-Oakenpants** [comment]
^^ yup, looked at that too. All good options. 1900 is one of the few hundreds gaps at the moment, and as you say is similar to a plugin - i.e optional extras added to the system. My only thought here is what about system addons that have no security/privacy etc implications - such as screenshots etc. Do we only list items here that fit our purpose, and put others under personal.

My thoughts are: it needs a round hundred, so it has room to grow over time - so 500, 600, 700 etc are all options. The section header would state that not all items have privacy/security/etc implications but are listed in this section to keep them all together. Anything that is not fit for our purpose, we either leave out, or we leave as inactive with a setup tag as a FYI. Add items in order they become system addons - make sure we add FF version numbers.

```js
/*** XXXX: SYSTEM ADDONS
     Blah blah
     [1] http blah blah
***/
/* xx01: disable experiments ***/
user_pref("experiments.enabled", false); //etc
/* xx02: disable Mozilla permission to silently opt you into tests ***/
user_pref("network.allow-experiments", false);

/* xx10: disable "Pocket"  (NEED FF version number)***/
user_pref("extensions.pocket.enabled", false); //etc
/* xx11: disable Activity Stream (FF54+) ***/
user_pref("browser.newtabpage.activity-stream.enabled", false);
/* xx12: disable Onboarding (FF56+) ***/
user_pref("browser.onboarding.enabled", false);
/* xx13: disable Screenshots (FF56+) ***/
   // user_pref("extensions.screenshots.disabled", true);

```

**Comment by Thorin-Oakenpants** [comment]
Info/ideas for section header:
- https://gecko.readthedocs.io/en/latest/toolkit/mozapps/extensions/addon-manager/SystemAddons.html
- > The set of default system add-ons are checked into mozilla-central under ./browser/extensions. These get placed into the features directory of the application directory at build time
   * check it out: https://dxr.mozilla.org/mozilla-central/source/browser/extensions
- Feel free to write up a section header guys and gals

**Comment by Thorin-Oakenpants** [comment]
clicktoplay-rollout: see [1365714 Release Flash plugin Click-to-Play (aka "Ask to Activate")](https://bugzilla.mozilla.org/show_bug.cgi?id=1365714) resolved fixed FF55 .. prefs to look for contain `ctprollout` AFAIK

**Comment by earthlng** [comment]
Thanks for the links, nice find! These 2 definitely need to go in the user.js.
From your first link [this](https://gecko.readthedocs.io/en/latest/toolkit/mozapps/extensions/addon-manager/SystemAddons.html#remove-all-system-add-on-updates) part is especially interesting... we can set `extensions.systemAddon.update.url` to a value to "remove all system add-on updates".
"updates" only though, not the pre-installed versions. But still, we should include that pref in the new section, somewhere at the top. We can set it to empty string (boring ;) or make it pretend there are no updates, or to "remove all system add-on updates".
I mean you have to give these guys credit, they really document their stuff very well and in great detail. Thx for that @mozillaGuys xD

> it needs a round hundred, so it has room to grow over time - so 500, 600, 700 etc are all options. 

600 is taken but I agree 500 or 700 are also good candidates.

> current sections

we should add that list to the Readme IMO.

> mortar and pdfjs are the other two not mentioned in earthlng's list

mortar doesn't seem to be a system addon. No idea what that is. pdfjs is likely just hosted there with the rest of them but will unfortunately probably not be made into a system-addon.

> clicktoplay-rollout: ... I think this is self contained - there are no prefs.

from looking at the [source](https://dxr.mozilla.org/mozilla-central/source/browser/extensions/clicktoplay-rollout) the addon consists of a single js file and it deals with `plugin.state.flash`.
Default value for that, at least in ESR52, is `2` (=enabled), while the addon will set it to `1` ([=click-to-play](https://dxr.mozilla.org/mozilla-central/source/dom/plugins/base/nsIPluginTag.idl#16))
I agree we can ignore this.

nits: `/* xx12: disable Activity Stream (FF54+) ***/` - Why FF54+? It's not included in FF54. Only pocket and screenshots are in FF54 (plus the 3 "ok" ones, aushelper, e10s + webcompat) [although webcompat probably deserves a closer look once it does something - I'm curious what it does]

**Comment by earthlng** [comment]
re: `clicktoplay` https://developer.mozilla.org/en-US/Firefox/Releases/55#Plugins
> For Firefox 55 **release version**, the plan is to activate this for 5% of users 2 weeks after release, 25% of users 4 weeks after release, and **100% of users 6 weeks after release**

FF56 will probably have `plugin.state.flash` defaulting to `1` then and the addon will get trashed

**Comment by earthlng** [comment]
[webcompat-reporter](https://dxr.mozilla.org/mozilla-release/source/browser/extensions/webcompat-reporter/bootstrap.js#17) - `extensions.webcompat-reporter.enabled`

**Comment by Thorin-Oakenpants** [comment]
> Activity Stream (FF54+) ***/ - Why FF54+?

Oh yee of little faith ... https://github.com/ghacksuserjs/ghacks-user.js/issues/87 - 2nd pref listed under NEW prefs for FF54. They (I assume) push these out as experiments: eg maybe 5% got this in 54, the pref was default false, btw - which makes sense as a master switch for giving consent - 54 is definitely when it landed, I can see it in my about:config at default value.

**Comment by Thorin-Oakenpants** [comment]
> we should add that list to the Readme IMO.

Really? Do me a PR so I can look at it. Make sure to use the /wikipiki/ images so the https://ghacksuserjs.github.io/ghacks-user.js/ as that url doesn't load these things -> ` :small_orange_diamond: ` . I would also strip off all the items in square brackets. What do you title this? Do you put it under usage (nah) or before (yeah!)? I don't really want to shove usage down too far, but the readme within the user.js now covers everything, so I don't mind. How do you stylize it: keep all caps (nah) or proper case (maybe) or capitalise first letter only (probably) - acronyms excepted.

**Comment by Thorin-Oakenpants** [comment]
Pocket - "Important: Remove the pocket icon from your toolbar first" - pretty sure this no longer applies since migration to system addon. I'll remove that line

**Comment by Thorin-Oakenpants** [comment]
Are we sure the two experiment prefs belong here. Are these used to push early "system addons"? I see no mention actual mention of this

But here's an [example](https://www.reddit.com/r/firefox/comments/6nbr1w/clarifying_some_things_about_the_thread_removed/) where an add-on was pushed. My question, is this used for system add-on trials, or is it a "normal" add-on eg containers and test-pilot experiments (I'm pretty sure both these shown in about:addons from memory). Aren't experiments normal and after the kinks are worked out migrate it to system? IDK

**Comment by Thorin-Oakenpants** [comment]
long way to go - just throwing this up here

```js
/*** 0500: SYSTEM ADD-ONS
     System add-ons are a method for shipping extensions, considered to be
     built-in features to Firefox, that are hidden from the about:addons UI
     [1] https://gecko.readthedocs.io/en/latest/toolkit/mozapps/extensions/addon-manager/SystemAddons.html
     [2] https://dxr.mozilla.org/mozilla-central/source/browser/extensions
***/
user_pref("ghacks_user.js.parrot", "xxxx syntax error: the parrot was killed by a system add-on!");
/* 0501: disable experiments
 * NEED-DESCRIPTION / CLARIFICATION
 * [1] https://wiki.mozilla.org/Telemetry/Experiments ***/
user_pref("experiments.enabled", false);
user_pref("experiments.manifest.uri", "");
user_pref("experiments.supported", false);
user_pref("experiments.activeExperiment", false);
/* 0502: disable Mozilla permission to silently opt you into tests
 * NEED-DESCRIPTION / CLARIFICATION ***/
user_pref("network.allow-experiments", false);

/* 0503: extensions.systemAddon.update.url ? ***/
// EARTHLNG: what does this achieve exactly - it could block updates to pdfjs

/* 0510: disable Web Compatibility Reporter
 * DO-DESCRIPTION & SCARYSTUFF ***/
user_pref("extensions.webcompat-reporter.enabled", false);
/* 0511: disable Pocket (FF39+?)
 * Pocket is a third party (now owned by Mozilla) "save for later" service
 * [1] https://en.wikipedia.org/wiki/Pocket_(application)
 * [2] https://www.gnu.gl/blog/Posts/multiple-vulnerabilities-in-pocket/ ***/
user_pref("extensions.pocket.enabled", false);
user_pref("extensions.pocket.api", "");
user_pref("extensions.pocket.site", "");
user_pref("extensions.pocket.oAuthConsumerKey", "");
/* 0512: disable FlyWeb (FF50+?)
 * Flyweb is a set of APIs for advertising and discovering local-area web servers
 * [1] https://wiki.mozilla.org/FlyWeb
 * [2] https://www.ghacks.net/2016/07/26/firefox-flyweb/ ***/
user_pref("dom.flyweb.enabled", false);
/* 0513: disable Shield (FF??+)
 * DO-DESCRIPTION & SCARYSTUFF
 * [1] https://wiki.mozilla.org/Firefox/Shield
 * [2] https://bugzilla.mozilla.org/show_bug.cgi?id=1308656 ***/
user_pref("extensions.shield-recipe-client.enabled", false);
/* 0514: disable Activity Stream (FF54+)
 * DO-DESCRIPTION & SCARYSTUFF
 * [1] https://wiki.mozilla.org/Firefox/Activity_Stream ***/
user_pref("browser.newtabpage.activity-stream.enabled", false);
/* 0515: disable Onboarding (FF55+)
 * DO-DESCRIPTION & SCARYSTUFF
 * [1] https://wiki.mozilla.org/Firefox/Onboarding ***/
user_pref("browser.onboarding.enabled", false);
/* 0516: disable AutoFormFill (FF55+)
 * DO-DESCRIPTION & SCARYSTUFF
 * [NOTE] heuristics controls Form Autofill on forms without @autocomplete attributes
 * [1]  ***/
user_pref("extensions.formautofill.experimental", false);
user_pref("extensions.formautofill.addresses.enabled", false);
user_pref("extensions.formautofill.heuristics.enabled", false);
user_pref("extensions.formautofill.loglevel", "Warn");
/* 0517: disable Screenshots (FF55+)
 * DO-DESCRIPTION & SCARYSTUFF
 * [1] https://github.com/mozilla-services/screenshots
 * [2] https://www.ghacks.net/2017/05/28/firefox-screenshots-integrated-in-firefox-nightly/ ***/
   // user_pref("extensions.screenshots.disabled", true);
```

**Comment by earthlng** [comment]
I'll address a couple things but then I really think we should take this step by step. Finish something then move on to the next and cleanup the comments first. Otherwise we end up with walls of text that are all over the place.

> Oh yee of little faith ...

Ok fair enough. If we go by when a pref first landed instead of when the SA is included for everyone in the FF release setup, that's fine. You can delete that comment of yours and I'll edit that part of my comment where I asked about that.

> Flyweb - I think was introduced in FF50

according to my diffs `dom.flyweb.enabled` was `new` in FF49, if that's what we are going by.
I have diffs for as far back as FF38vs39 - I should probably upload them to my repo so you or anyone else can use them as well for things like this.

> /* 0517: disable Screenshots (FF55+)

Screenshots definitely exists already in FF54. ie it's shipped with the setup for everyone.
`pref("extensions.screenshots.system-disabled", true);` was `new` in FF54. `pref("extensions.screenshots.disabled", false);` is `new` in FF55b1

---

I don't think we necessarily need `DESCRIPTION & SCARYSTUFF`. A couple of links are good enough IMO.

> /* 0503: extensions.systemAddon.update.url ? ***/
// EARTHLNG: what does this achieve exactly - it could block updates to pdfjs

It could, idk. If we don't "disable" this url then it could install updated versions of addons that someone completely removed from the /browser/features/ folder. Just saying. If we don't care about that, fine.

**Comment by earthlng** [comment]
> Are we sure the two experiment prefs belong here. ...

We can name the section "Mozilla System-Addons / Experiments"

>  Isn't Test Pilot how they now experiment?

Test Pilot, afaik, is something that a user has to install. The other stuff (experiments, shield-shit, etc) is all done without user consent.

**Comment by earthlng** [comment]
> If you can correct me, awesome. Maybe go find a bugzilla source. Otherwise I'm going with FF39+

Yes, according to my diffs the old, original `browser.pocket.*` prefs were `new` in FF39.
In FF46 they added `pref("extensions.pocket.enabled", true);` and removed all the `browser.pocket.*` prefs. All the other replacement prefs (`extensions.pocket.*`) apart from `extensions.pocket.enabled` are now set by the addon itself. (`.oAuthConsumerKey`, `.api`, etc)
IMO we can remove the following prefs because `.enabled` is the master switch:
```js
user_pref("extensions.pocket.api", "");
user_pref("extensions.pocket.site", "");
user_pref("extensions.pocket.oAuthConsumerKey", "");
```
Proof: [here](https://dxr.mozilla.org/mozilla-central/source/browser/extensions/pocket/bootstrap.js#476) you can see that in the addon's startup() function it sets the default prefs (but doesn't overwrite them if already set), then migrates the old prefs, adds an observer (so that toggling the pref in about:config actually does something) and then returns if the master switch is set to false ie it never runs `PocketOverlay.startup()`
oAuthConsumerKey also seems to be the same for everyone anyway.

**Comment by Thorin-Oakenpants** [comment]
- Would be nice to know if "experiments" was just additional telemetry stuff but ya know what - testing out new features is telemetry and since system addons are now a thing, this is probably themechanism they use, so it's all tied together anyway - forget I mentioned it. Test Pilot is just another way for them to try new things I suppose, and yeah, you have to install it, and it shows in about:addons
- Section renamed in list above and draft (excellent idea)
- Screenshots: I was aware there were two prefs, in my haste to throw up a sample I just added one. fixed
- Pocket in 39, sweet - done (thanks for confirming). Note when I do the commit I will adjust the one we already have in deprecated FF46 to match numbers (yup aware of the pref renaming) - already had it covered
- Flyweb in 49, sweet - done (thanks for confirming)
- `extensions.systemAddon.update.url` - I do not think we should list this or meddle with this. You either physically remove an item/xpi or use the built in switches. There's not much that can't be disabled by the looks of it (despite the statement in link [[1]](https://gecko.readthedocs.io/en/latest/toolkit/mozapps/extensions/addon-manager/SystemAddons.html)  that "[they] cannot be user disabled" . Besides, these could contain fixes for security or privacy. The whole point was the quick updating, restartlessness. Seems Monty Pythonesque to turn that off (might as well buy a dead parrot to begin with).
- `DESCRIPTION` (& scarystuff). Well I wanted every system addon to have a link to official mozilla page, a simple description of WTF it is, and to say WHAT it bad about it as to why we disable (because not all of these are "bad" - eg screenshots, AFAIK - just don't use the online bit I guess). That was all. Maybe add a line to the section description?

**Comment by earthlng** [comment]
`extensions.systemAddon.update.url`

> I do not think we should list this or meddle with this.

I disagree. We should at least list it as inactive.

> You either physically remove an item/xpi or use the built in switches.

But that's the thing - even if you remove an xpi completely they can still install you a new version

> There's not much that can't be disabled by the looks of it (despite the statement in link [1] that "[they] cannot be user disabled")

Unless you remove them from your install-dir they are "technically" impossible to disable ie they show up in about:support as "enabled". Yes you can "stop" most of them from doing anything but that's not what [they] are talking about.

> Besides, these could contain fixes for security or privacy. The whole point was the quick updating, restartlessness. Seems Monty Pythonesque to turn that off (might as well buy a dead parrot to begin with).

AFAIK if that's a concern we also need to enable auto-install "app-updates" according to [this](https://bugzilla.mozilla.org/show_bug.cgi?id=1335224#c14) comment. IDK if that's still the case though but I've certainly never received any system-addon updates or fresh installs (fe the HSTS one they mention in that comment)

**Comment by Thorin-Oakenpants** [comment]
(That ticket opens a can of worms). APP updates are **expected** on a regular basis (8 weekly) and you get notifications. I do not know exactly how a system addon update works (outside changes bundled with an app update), but I would suspect that they are irregular and could be almost never. The whole purpose is restartless push to fix issues (without doing minor releases) and this IMO outweighs anything else - eg imagine an exploit fix to pdfjs, or a privacy leak fix in Pocket, or a security fix in AutoFormFill.

What I'm concerned about now is that with out default settings - i.e check but allow me to update the application - that as per that ticket, system addons will not update

**Comment by earthlng** [comment]
I'd replace link [1] for FlyWeb with https://flyweb.github.io/ (which links to [1] right at the top). And maybe add another link to https://wiki.mozilla.org/FlyWeb/Security_scenarios

**Comment by earthlng** [comment]
> What I'm concerned about now is that with our default settings - i.e check but allow me to choose when to update the application - that as per that ticket, system addons will not update

that's what I meant

**Comment by Thorin-Oakenpants** [comment]
* 0513: disable Shield (FF??+)  <- What version was this? I think it was first quarter this year, so recent
 * [2] https://bugzilla.mozilla.org/show_bug.cgi?id=1308656 ***/ <--DROP THIS LINK? overkill, it adds nothing IMO

**Comment by Thorin-Oakenpants** [comment]
Re updating system addons with our settings - then the solution is to log a bug. Not sure how we can test - maybe open an issue with one of the system addon github repos and ask for help?

**Comment by earthlng** [comment]
* Shield - this is a bitch!
https://wiki.mozilla.org/Firefox/Shield
>  December 2016 - [SHIELD System Add-on lands in mozilla-central](https://bugzilla.mozilla.org/show_bug.cgi?id=1308656)

according to the [changeset](https://hg.mozilla.org/mozilla-central/rev/b35d6cb8df5d) for that bugzilla shield was first deployed in FF53

The only good thing about Shield is that it lead the removal of SelfSupport from FF's core:
https://hg.mozilla.org/mozilla-central/rev/5a8f2dcbeac0
The master pref for Shield seems to be somewhat of a red herring. I can't find the code where that is actually used. It also still looks at `browser.selfsupport.enabled` but it needs that pref to be `false` for Shield to do something.
The problem with this addon is that it auto-uninstalls when it's done and I'm not sure we should try to stop it from gathering telemetry ie hit the return [here](https://github.com/mozilla/normandy/blob/master/recipe-client-addon/lib/ShieldRecipeClient.jsm#L72). Instead we should probably just neutralize the url where it sends its data to by setting `.api_url` to empty string. And also include `extensions.shield-recipe-client.enabled` just in case it does something somewhere.

**Comment by Thorin-Oakenpants** [comment]
I guessed 53 from when they closed the ticket as fixed 6 months ago, and allowed for a couple of versions out from stable, and also from the recent reddits and here and at PK's about the pref suddenly turning up starting about 53 but before 54 :)

https://dxr.mozilla.org/mozilla-central/search?q=shield-recipe-client.api&redirect=false
I do not understand this `.api_url` - it doesn't seem to exist except in tests, or it sets it, or it logs it. I don't follow.

**Comment by earthlng** [comment]
> I do not understand this .api_url

It all starts [here](https://github.com/mozilla/normandy/blob/master/recipe-client-addon/bootstrap.js#L16) in bootstrap.js function startup(). That's the entry point for all older addons. There it calls `ShieldRecipeClient.startup()` which is [here](https://github.com/mozilla/normandy/blob/master/recipe-client-addon/lib/ShieldRecipeClient.jsm#L55) On the first line it sets the default values which you can see on line 34. The api_url is then later, maybe among other places, checked [here](https://github.com/mozilla/normandy/blob/master/recipe-client-addon/lib/RecipeRunner.jsm#L73) to make sure it's a correct url.
Granted at that point we would have already returned if either toolkit.telemetry.unified or extensions.shield-recipe-client.enabled are false, and we probably don't need to include api_url in the user.js

**Comment by Thorin-Oakenpants** [comment]
> we probably don't need to include api_url in the user.js

^^ that's what I wanted to here :smiley: :smile: :)

**Comment by earthlng** [comment]
* You don't want to include api_url for shield but keep the ones for pocket?? why?
* FlyWeb is done, right? I ticked it off in the first comment list.
* webcompat-reporter should not be 0510, it's maybe coming in FF56 or even later.
If they also add the master switch pref to FFs preferences files, as they did with all the other ones, we will see it in the FF56b1 diff

**Comment by Thorin-Oakenpants** [comment]
> You don't want to include api_url for shield but keep the ones for pocket?? why?

IDK. All that pocket stuff went thru several stages, plus I have had the pocket system addons, but I have yet to see a shield one. I thought the shield api pref didn't exist. Maybe we can clear out the pocket ones and just keep `extensions.pocket.enabled`

**Comment by Thorin-Oakenpants** [comment]
- FYI [1308981 separate system add-ons from extensions in about:support](https://bugzilla.mozilla.org/show_bug.cgi?id=1308981) - resolved fixed FF55

We could add this to info to the section header. eg "To view your system add-ons go to about:support, they are listed under `Features`" (I do not have 55, can anyone tell me the section title?)

**Comment by earthlng** [comment]
In nightly 55+56 they are listed under "Nightly Features" and the normal addons are listed under "Extensions".

**Comment by Thorin-Oakenpants** [comment]
Thanks. Once 55 to lands we should confirm the title. I'll add it without the word Nightly for now

**Comment by Thorin-Oakenpants** [comment]
Added `user_pref("extensions.shield-recipe-client.api_url", "");`

**Comment by Thorin-Oakenpants** [comment]
Added some stuff - see revamped js up above

* Shield is an telemetry system (including Heartbeat) that can also push and test "recipes"
* Activity Stream replaces "New Tab" with a new design based on metadata and browsing behavior, and includes telemetry as well as web content such as snippets and "spotlight"
* [2] https://www.ghacks.net/2016/02/15/firefox-mockups-show-activity-stream-new-tab-page-and-share-updates/

I ticked off Activity Stream, Shield, Pocket in the first post, as well as e10s which will become obsolete

ToDo
- Onboarding: add a one-liner about what it is: eg "Onboarding is a new user/profile tour experience"
- Screenshots: check out the diff between the two prefs, and check out about shots containing data and cloud services
- AutoFormFill: links and one-liner maybe about local collection of types of data and exposure
- Web Compatibility Reporter: one-liner? link?

Other:
- aushelper & webcompat (nothing more we can do with AFAIK)
- follow-on-search
- presentation

**Comment by earthlng** [comment]
* aushelper - this is fine. It changes the app update url if someone has Websense installed because that apparently causes problems with certain FF versions. And also something to do with CPUs, idk exactly. I think "aus" stands for application update service or something like that.

I looked at some of the addons on DXR (nightly) to see what they do and if there are any other prefs for them...
* webcompat - the current version on DXR includes 1 pref `extensions.webcompat.perform_ua_overrides` but it gets reset if it's user-set right at the beginning in the startup function. Atm (in nightly!) that's all it does: user-agent overrides. But so far there's only [one override](https://dxr.mozilla.org/mozilla-release/source/browser/extensions/webcompat/content/data/ua_overrides.jsm#41) setup, for an example site to test if the addon works.
```js
  /*
   * This is a dummy override that applies a Chrome UA to a dummy site that
   * blocks all browsers but Chrome.
   *
   * This was only put in place to allow QA to test this system addon on an
   * actual site, since we were not able to find a proper override in time.
   */
```
It *does* register an observer for that pref so you can toggle it off if the override causes problems, but the only way to completely stop this addon from the user.js is to use "pref()" instead of "user_pref" (or lock the pref). But I don't think we should do that. We *can* add the pref commented out IMO.

**Comment by Thorin-Oakenpants** [comment]
removed those 3 pocket prefs leaving just the master switch

**Comment by Thorin-Oakenpants** [comment]
so we don't have to scroll heaps

part 2 
```js
/* 0514: disable Screenshots (FF54+)
 * [1] https://github.com/mozilla-services/screenshots
 * [2] https://www.ghacks.net/2017/05/28/firefox-screenshots-integrated-in-firefox-nightly/ ***/
   // user_pref("extensions.screenshots.system-disabled", true); // (FF54+)
   // user_pref("extensions.screenshots.disabled", true); // (FF55+)
/* 0515: disable Onboarding (FF55+)
 * [1] https://wiki.mozilla.org/Firefox/Onboarding ***/
user_pref("browser.onboarding.enabled", false);
/* 0516: disable AutoFormFill (FF55+)
 * [NOTE] heuristics controls Form Autofill on forms without @autocomplete attributes
 * [1]  ***/
user_pref("extensions.formautofill.experimental", false);
user_pref("extensions.formautofill.addresses.enabled", false);
user_pref("extensions.formautofill.heuristics.enabled", false);
user_pref("extensions.formautofill.loglevel", "Warn");
/* 0517: disable Web Compatibility Reporter ***/
user_pref("extensions.webcompat-reporter.enabled", false);
```

Need: A single line what onboarding is. A link for AutoFormFill. Check `formfill.loglevel` (what levels exist). Explain what Web Compat Reporter does. Make distinction between the two Screenshot prefs.

Anyone wanna build a draft, man :snowman:

**Comment by Thorin-Oakenpants** [comment]
Screenshots: https://reviewboard.mozilla.org/r/145308/diff/2#index_header - We can remove `extensions.screenshots.system-disabled` IMO Y/N?

Onboarding: https://dxr.mozilla.org/mozilla-central/source/browser/extensions/onboarding/README.md#18 .. I assume this is only if the tour hasn't been completed. BTW, part of the onboarding includes automigration of bookmarks from the default browser. Also, they can bump version numbers and force new tours, so it's not really just new installs (I think). And then there's this https://dxr.mozilla.org/mozilla-central/source/browser/extensions/onboarding/content/onboarding.js#310 - hidden tabs. Going with this
```js
/* 0515: disable Onboarding (FF55+)
 * Onboarding is an install/new profile experience/user-tour. Everytime `about:home` or
 * `about:newtab` page is opened, the onboarding overlay is injected into that page
```

-----

AutoFormFill
- Should look at `dom.forms.autocomplete.experimental` (not sure when it landed) as this is used in the autoformfill tests.
- Data is stored in a JSON file (that's not very secure!) - https://dxr.mozilla.org/mozilla-central/source/browser/extensions/formautofill/ProfileStorage.jsm
- There is a UI pref - what is it (if you have 55+ let me know if it has changed) - https://dxr.mozilla.org/mozilla-central/source/browser/extensions/formautofill/FormAutofillPreferences.jsm -  It should be [SETTING] Options>Privacy>Forms & Passwords>Enable Profile Autofill
```js
      formAutofillGroup.hidden = document.location.href != "about:preferences#privacy";
      // Use .setAttribute because HTMLElement.dataset is not available on XUL elements
      formAutofillGroup.setAttribute("data-category", "panePrivacy");
      formAutofillGroup.appendChild(caption);
      caption.appendChild(captionLabel);
      captionLabel.textContent = this.bundle.GetStringFromName("preferenceGroupTitle");
```
- Links
  * [1] https://wiki.mozilla.org/Firefox/Features/Form_Autofill
  * [2] https://www.ghacks.net/2017/05/24/firefoxs-new-form-autofill-is-awesome/

---

Sidenote: reading about so many of these system add-ons, I see an awful lot of tickets references to telemetry - A lot of these are/were `experimental`, and some telemetry will now not be needed as they land in stable - but it really makes me question if these system add-ons RESPECT the telemetry opt-out. We should follow up on that.

**Comment by earthlng** [comment]
> ~~Anyone wanna build a draft, man~~ :snowman:  I guess not! Might as well do it myself then *sigh*

yeah sorry mate, was kinda busy with the canvas blocker testing stuff 🤦‍♂️ 
offtopic but how the fuck is `:man_facepalming:` = a woman facepalming ??? gender madness on github now too?! well that's just fucking great /rant

> AutoFormFill
> * Data is stored in a JSON file (that's not very secure!)

They also want to add support for credit cards 🤦‍♂️ ez matching email addresses + cc numbers - nice.
https://blog.nightly.mozilla.org/2017/05/23/preview-form-autofill-in-firefox-nightly/

> Web Compatibility Reporter ... basically enables a Report Site Issue button which sends a screenshot to Mozilla (and maybe you can enter some text who knows).

That's in a test file and I didn't look at those. But you're right there's something in it that does that (I looked at the code that came bundled with FF56 nightly). There's another pref that holds the URL for the endpoint: https://dxr.mozilla.org/mozilla-central/source/modules/libpref/init/all.js#4939
You can look at the site - it's already up and has a nice little window where the screenshot would go.

**Comment by earthlng** [comment]
> We can remove `extensions.screenshots.system-disabled` IMO Y/N?

yes

> `extensions.formautofill.loglevel` - IMO we can drop this pref, since if the addon is disabled, it wont log anything

definitely. Even if it did log something, these .loglevel prefs usually only control the verbosity of **console output**

**Comment by earthlng** [comment]
WTF?! now the woman facepalming turned back into a man?!! am I the only one seeing this? WTH is happening??

**Comment by Thorin-Oakenpants** [comment]
:man_facepalming: :woman_facepalming: i see gender people

Just type `:woman_` and `:man_` and there's a heap of them. What next? `:man_ballscratching:` and `:woman_ballscratching:` (no one said it had to be **HER** balls) .. so I could totally expect that

**Comment by Thorin-Oakenpants** [comment]
Onboarding is a ~~[I do not know if it will be triggered for existing profiles, I do not know if it will be triggered for installs to be run at least once especially if FF is not the default, I do not know if it will be triggered by pushed new versions to push a Mozilla incentive interactive tour for some new feature, I do not know if it will be triggered for a new profile but I suspect it WILL be since the prefs will be a default and it will want to run]~~ pile of interactive bullshittery triggered on unsuspecting users

What can you come up with?

**Comment by Thorin-Oakenpants** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=1369046 - its a button you can click, you know, like on your toolbar. The test is just a test

**Comment by earthlng** [comment]
> Onboarding is a pile of interactive bullshittery triggered on unsuspecting users

`:woman_laughing:` ["to interact with unbiased subsets of Firefox users"](https://github.com/mozilla/normandy) please! A little bit of PC is not too much to ask now is it? I mean ... you see gender *sigh* I shouldn't be surprised xD `:woman_shaking_head:`

kidding aside I think your statement sums it up pretty accurately. Why try to sugarcoat the ugly truth

**Comment by Thorin-Oakenpants** [comment]
```js
 * Onboarding is an interactive tour/setup for new installs/profiles and features. Every time
 * about:home or about:newtab is opened, the onboarding overlay is injected into that page
```
That sums it up. They can trigger it for new features if they want to, at any time - they could push a nasty 12 page next next next slideshow of crap on anyone at any time. But its purpose seems to be for `new` installs / profiles

**Comment by earthlng** [comment]
Yeah I mean we can always change it later. Someone #cough needs to take one for the team and test this shit I guess. #cough #notme #cough - oh sorry :woman_cough:

**Comment by earthlng** [comment]
`follow-on-search`
https://blog.nightly.mozilla.org/2017/07/19/these-weeks-in-firefox-issue-20/
> The Follow On Search Count addon was [**shipped to 54 population**](https://bugzilla.mozilla.org/show_bug.cgi?id=1372307).

There's currently no pref to stop this addon (except setting `distribution.id` to a non-empty string). Atm it only collects telemetry and that's stopped by disabling telemetry. There is however a logging pref `extensions.followonsearch.logging` that you can set to `true` to get console output about what it *would* send as telemetry. IMO we should include that pref (inactive) so we don't forget that this thing exists, and then check back from time to time to see if it "evolved" and if by then there are prefs to disable it completely.

**Comment by Thorin-Oakenpants** [comment]
so we don't have to scroll heaps

part 3
```js
/* 0514: disable Follow On Search (FF54+) ***/
/* 0517: disable AutoFormFill (FF55+)
 * [SETTING] Options>Privacy>Forms & Passwords>Enable Profile Autofill
 * [NOTE] Stored data is NOT secure (uses a JSON file)
 * [NOTE] Heuristics controls Form Autofill on forms without @autocomplete attributes
 * [1] https://wiki.mozilla.org/Firefox/Features/Form_Autofill
 * [2] https://www.ghacks.net/2017/05/24/firefoxs-new-form-autofill-is-awesome/ ***/
user_pref("extensions.formautofill.addresses.enabled", false);
user_pref("extensions.formautofill.experimental", false);
user_pref("extensions.formautofill.heuristics.enabled", false);
user_pref("dom.forms.autocomplete.experimental", false) // (FF52+)
``

Again .. not ~~happy~~ sure with `dom.forms.autocomplete.experimental` being in here. It was introduced in 52 and clearly uses a different naming convention, yet it seems tied. Please 'splain. And  feel free to fill in 0514

**Comment by earthlng** [comment]
https://blog.mozilla.org/data/2017/06/05/measuring-search-in-firefox/ 

I can't stand this fucking marketing talk. I'm pretty sure they will eventually intercept the follow-on-search requests and add their utm tracking parameters.

> This data will allow us to **greatly** improve the Firefox search experience while still respecting user privacy.

"greatly" huh? bullshit! They hadn't even started collecting anything at that point, how could they already know the outcome?! and not just "improve" either but "greatly improve". BS!

"Search is blah blah **how Mozilla makes much of its revenue.**" = this is the only time where they come right out and say what all of this is really about.

>Firefox telemetry already includes a count of the searches users perform in all Firefox search bars. Firefox does not yet count follow-on searches.

= we already make money from searches users perform and we want money for follow-on searches as well.

> When you search with one of the search engines that we include in Firefox, we will increment a counter for each follow-on search.

= they can only make money from SEs included in Firefox, basically admitting that it's not really about improving " Firefox search experience for our users", because they can also get telemetry for user-added search engines. Why are they not interested in those?

> We seek these new measurements to **gain** missing insight into a crucial browser interaction.

= "gain" is an interesting choice of words. It hints at what they really had in mind when writing this. There are a few more of these: "our limited resources", "continue to sustain"

read the blog post again and tell me I'm wrong.

**Comment by Thorin-Oakenpants** [comment]
> And feel free to fill in 0514

**Comment by Thorin-Oakenpants** [comment]
What exactly is a "follow-on search"?

**Comment by Thorin-Oakenpants** [comment]
Why comment out // user_pref("extensions.followonsearch.logging", true) ? Why is it true? Needs more info. eg "Enable the logging pref to see in the console what is being collected."?

We could add some lines to the section header where these are kept and how to kill: I for one am going to remove the ones I do not want anything to do with eg

```js
/***
Some system add-ons have no on-off preferences. Instead you can to manually remove them.
However, app updates will restore them. Additionally, they may be updated/restored automatically (see 0505)
   * Portable: `...\App\Firefox64\browser\features\` (or `App\Firefox\etc` for 32bit)
   * Windows: `...\Program Files\Mozilla\?????` (or `Program Files (X86)\etc` for 32bit)
***/
```

```js
/* 0505: block URL used for system add-on updates
   [NOTE] Not recommended as you will not get security/privacy fixes until you update Firefox ***/
   // user_pref("extensions.systemAddon.update.url", "");
```

I'm not sure on that update url. If you delete it, will it restore it?

**Comment by Thorin-Oakenpants** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=863246#c154

Another reason Onboarding sucks - injecting code into pages and opening up resources to be accessible

**Comment by earthlng** [comment]
`extensions.systemAddon.update.url` was added in FF44

> Again, I ask, in any of the code (for these system addons) is there a lookup to respect a user's opt-out of telemetry

The shield client for example looks at the telemetry pref before initiating the recipe manager.
But unless an addon uses google analytics, telemetry always uses the built-in telemetry framework. If that's disabled, afaik it doesn't matter whether there are telemetry gathering calls made in the addon.

> Windows: `...\Program Files\Mozilla\?????`

...\Program Files\Mozilla Firefox\browser\features\

> I'm not sure on that update url. If you delete it, will it restore it?

will what restore what?

**Comment by Thorin-Oakenpants** [comment]
> will what restore what?

If you delete a system addon, will the update check restore it - or will it only update those that are present

**Comment by earthlng** [comment]
> `extensions.systemAddon.update.url` was added in FF44

btw I've uploaded all my diffs for as far back as FF38vs39.

https://github.com/earthlng/FFprefs-diffs/search?q=extensions.systemAddon.update.url

**Comment by Thorin-Oakenpants** [comment]
I used to have nilla FF's going back to 20 (and even some 3.x and earlier versions) - but I nixed em all

**Comment by Thorin-Oakenpants** [comment]
I think we can tick off aushelper and webcompat

**Comment by earthlng** [comment]
> If you delete a system addon, will the update check restore it - or will it only update those that are present

It asks for updates based on your FF version. f.e lets say screenshots_v1.0 comes bundled with FF54:
if you remove the addon it will probably not get "restored" as long as there's no new version. It will probably re-install (into your profile folder!) a new version (v1.1) regardless if you removed v1.0 or not, because otherwise they could never deploy completely new addons

**Comment by Thorin-Oakenpants** [comment]
presentation: https://dxr.mozilla.org/mozilla-central/source/browser/extensions/presentation/content/PresentationDevicePrompt.jsm#8 

> This is the implementation of nsIPresentationDevicePrompt XPCOM.
> * It will be registered into a XPCOM component by Presentation.jsm.
>  This component will prompt a device selection UI for users to choose which
> * devices they want to connect, when PresentationRequest is started.

--> https://dxr.mozilla.org/mozilla-central/source/accessible/jsat/Presentation.jsm

Which is about braille and other accessibility thingies

I think we can ignore this. We already have 2513 which disables the Presentation API

**Comment by earthlng** [comment]
> Again .. not happy sure with `dom.forms.autocomplete.experimental` being in here. It was introduced in 52 and clearly uses a different naming convention, yet it seems tied. Please 'splain

where did you get this from? it's not currently in the user.js, did you delete it by mistake or what?

**Comment by Thorin-Oakenpants** [comment]
`dom.forms.autocomplete.experimental` has never been in the user js (search the repo, I did). I came across it in the autoformfill tickets - landed FF52 - see https://reviewboard.mozilla.org/r/85784/diff/3#index_header

I may be wrong on that - that bew code is just using the pref. What do your diffs say for when it arrived (I'd look but I'm overworked).

**Comment by Thorin-Oakenpants** [comment]
FYI: shield recipe for measuring resistFingerprinting etc - https://bugzilla.mozilla.org/show_bug.cgi?id=1356381

**Comment by Thorin-Oakenpants** [comment]
@grauenwolfe Can you please tell me where on a Mac are system add-ons stored?

FYI
* Portable
   * "...\App\Firefox64\browser\features\" (or "App\Firefox\etc" for 32bit)
* Windows
   * "...\Program Files\Mozilla\browser\features" (or "Program Files (X86)\etc" for 32bit)

At the very least you should have an `aushelper@mozilla.org.xpi` file (in case you want to search for it)

**Comment by Thorin-Oakenpants** [comment]
@earthlng 
I have edited the first post checklist, speak up if you're unhappy. Down to one item. Suggest this, and then I can commit, close the issue, and get Martin to do an article. :+1: if you're ok with it

```js
/* 0513: disable Follow On Search (FF53+)
 * Just DELETE the XPI file in your system add-ons directory
 * [1] https://blog.mozilla.org/data/2017/06/05/measuring-search-in-firefox/ ***/
```

**Comment by Thorin-Oakenpants** [comment]
FYI: https://addons.mozilla.org/en-US/firefox/addon/firefox-pioneer/
https://medium.com/firefox-context-graph/make-firefox-better-with-pioneer-10c82d0f9301

And there are other experiments. So glad we have a system add-on section now

**Comment by Atavic** [comment]
[Shield Recipe Client](https://github.com/ghacksuserjs/ghacks-user.js/issues/116#issuecomment-315542875)

**Comment by Thorin-Oakenpants** [comment]
Article: https://www.ghacks.net/2017/07/30/control-firefox-system-add-ons-with-the-ghacks-user-js-file/

I sent Martin the the entire 0500 section code a day ago, along with links this to this issue and the one for the system add-ons update bug. I guess he felt he only wanted to mention a handful of prefs. Nothing about Onboarding, Screenshots, Auto Formfill etc, none of our links

:sad face:

**Comment by earthlng** [comment]
we missed `extensions.webcompat.perform_ua_overrides`

**Comment by Thorin-Oakenpants** [comment]
- https://bug98304.bugzilla.mozilla.org/show_bug.cgi?id=1307792
- https://github.com/mozilla/webcompat-addon/pull/4/files

And?

**Comment by earthlng** [comment]
And what? And what are those links supposed to tell me?

https://github.com/ghacksuserjs/ghacks-user.js/issues/171#issuecomment-316096653

we can ignore that pref IMO. I saw it in @Theemim's diff and thought we forgot to add it, nvm.
